File: Case1/compute_2nd_ANOVA.m
function [Err_2nd, C_O0_t, C_O1_t, Pairs, Pair_all] = compute_2nd_ANOVA(cur_path,Pair_all, iPair,C_O0_t, C_O1_t, A2, Pairs, iO, C_O0, C_O1, kl_term,Node_Num, C_P0, C_P1,C_P2,tt,Obs_Node,Obs_Num)
                                                                       
% compute the 2nd order ANOVA and estimate the error of estimation of Var(Output(j))
global Npair C_O2
    newPair = 1;
    for i=1:Npair
        if max(abs(Pair_all(iPair,3:4)-Pairs(:,i)'))==0
            newPair = 0;
            break
        end
    end
    
    if newPair
        Npair=Npair+1;
    
        P = [-sqrt(3); 0; sqrt(3)]; % 1D collocation points 

        % right hand side vector b
        
        CP = zeros(kl_term,1); CP(Pair_all(iPair,3))=P(1); CP(Pair_all(iPair,4))=P(1);
        CP1(:,1) = CP;
        CP = zeros(kl_term,1); CP(Pair_all(iPair,3))=P(3); CP(Pair_all(iPair,4))=P(1);
        CP1(:,2) = CP;
        CP = zeros(kl_term,1); CP(Pair_all(iPair,3))=P(1); CP(Pair_all(iPair,4))=P(3);
        CP1(:,3) = CP;
        CP = zeros(kl_term,1); CP(Pair_all(iPair,3))=P(3); CP(Pair_all(iPair,4))=P(3);
        CP1(:,4) = CP;
        save 'CP1.mat' CP1;
        save 'C_O0.mat' C_O0
        save 'C_O1.mat' C_O1
        
        D=zeros(4,Obs_Num);
        for i = 1:4
            fprintf('2_nd_%d_%d\n',iO,i);
%             cd(strcat(cur_path,'\start\start (',num2str(i),')'))
            p= MapParameters(CP1(:,i),kl_term,C_P0,C_P1,C_P2);
%             D(i,:) = (Hydrus(p,Node_Num,tt,Obs_Node)-ANOVA(CP1(:,i),C_O0,C_O1,kl_term))';
            zz=ogs_call(p,i,tt);
            D(i,:)=zz-ANOVA(CP1(:,i),C_O0,C_O1,kl_term);             
        end
        b = zeros(9,Obs_Num);
        b(1,:) = D(1,:);
        b(3,:) = D(2,:);
        b(7,:) = D(3,:);
        b(9,:) = D(4,:);
         
        k=A2\b;
        for s =1:Obs_Num
            t=reshape(k(:,s),3,3);
            C_O0_t(s) = C_O0_t(s)+t(1,1);
            C_O1_t(s,:,Pair_all(iPair,4)) = C_O1_t(s,:,Pair_all(iPair,4))+t(1,2:3);
            C_O1_t(s,:,Pair_all(iPair,3)) = C_O1_t(s,:,Pair_all(iPair,3))+t(2:3,1)';
            C_O2(s,:,:,Pair_all(iPair,3),Pair_all(iPair,4)) = t(2:3,2:3);
            if s == iO
                Pair_all(iPair,2)=sum(sum(t(2:3,2:3).^2));
            end
        end
        
        % Ratio = mean(Pair_all(1:iPair,2)./(Pair_all(1:iPair,1)+(1e-10)));
        Ratio= sum(Pair_all(1:iPair,2))/(sum(Pair_all(1:iPair,1))+(1e-10));
        Err_2nd = Ratio*sum(Pair_all(iPair+1:end,1));
        Pairs = [Pairs,Pair_all(iPair,3:4)'];

    else
        % if not new pair
        Pair_all(iPair,2)=sum(reshape(C_O2(iO,:,:,Pair_all(iPair,3),Pair_all(iPair,4)).^2,4,1));
        % Ratio = mean(Pair_all(1:iPair,2)./(Pair_all(1:iPair,1)+(1e-10)));
        Ratio= sum(Pair_all(1:iPair,2))/(sum(Pair_all(1:iPair,1))+(1e-10));
        Err_2nd = Ratio*sum(Pair_all(iPair+1:end,1));
    end

File: Case2/utils/compute_2nd_ANOVA.py
# -*- coding: utf-8 -*-

import numpy as np
import os
from ANOVA_parallel import ANOVA_parallel
from write_values import write_ki
from read_values import read_obs,read_obs_alltime
from runexe import runexe
from para_key_mod import para_keys_modify
from MapParameters import MapParameters
from time_modify import time_modify
def compute_2nd_ANOVA(pair_all,iPair,C_O0_t,C_O1_t,A2,Pairs,iO,C_O0,C_O1,kl_term,out_node,Npair,C_O2,C_P0,C_P1,C_P2,ki_mean,root_directory,Nod_num,filename_KI,obs_Num,filename_result,filename_time,tt):
    newPair=1
    for i in xrange(Npair):
        if np.max(np.abs(pair_all[iPair,2:4]-Pairs[:,i].T))==0:
            newPair=0
            break
    CP1=np.zeros((kl_term,4))
    if newPair:
        Npair=Npair+1
        P=np.array([-np.sqrt(3),0,np.sqrt(3)])
        CP=np.zeros(kl_term)
        CP[int(pair_all[iPair,2])]=P[0]
        CP[int(pair_all[iPair,3])]=P[0]
        CP1[:,0]=CP
        
        CP=np.zeros(kl_term)
        CP[int(pair_all[iPair,2])]=P[2]
        CP[int(pair_all[iPair,3])]=P[0]
        CP1[:,1]=CP
        
        CP=np.zeros(kl_term)
        CP[int(pair_all[iPair,2])]=P[0]
        CP[int(pair_all[iPair,3])]=P[2]
        CP1[:,2]=CP
        
        CP=np.zeros(kl_term)
        CP[int(pair_all[iPair,2])]=P[2]
        CP[int(pair_all[iPair,3])]=P[2]
        CP1[:,3]=CP
        
        temp=np.zeros((4,out_node))
        for i in range(4):
#            y=np.squeeze(np.log(ki_mean)+np.dot(Fi[:,:kl_term],CP1[:,i]))
#            y=np.exp(y)
            p=MapParameters(CP1[:,i],kl_term,C_P0,C_P1,C_P2)
            np.savetxt(r'../para_updating_{0}_{1}_22.txt'.format(tt,i),p)
            p=np.exp(p)
            p=para_keys_modify(p)
            write_ki(Nod_num,p,i,filename_KI,root_directory)
            time_modify(tt,i,filename_time,root_directory)
            # if np.mod(i,4)==0:
            print '2_nd_{0}_{1}'.format(iO,i)
            runexe(i,root_directory)
            
            obs_p2=read_obs_alltime(tt,0,obs_Num,Nod_num,root_directory,filename_result)
            temp[i,:]=obs_p2-ANOVA_parallel(CP1[:,i],C_O0,C_O1,kl_term)
        
        b=np.zeros((9,out_node))
        b[0,:]=temp[0,:]
        b[2,:]=temp[1,:]
        b[6,:]=temp[2,:]
        b[8,:]=temp[3,:]
        
        k=np.dot(np.linalg.inv(A2),b)
        for s in range(out_node):
            t=k[:,s].reshape((3,3))
            C_O0_t[s]=C_O0_t[s]+t[0,0]
            C_O1_t[s,:,int(pair_all[iPair,3])]=C_O1_t[s,:,int(pair_all[iPair,3])]+t[0,1:3]
            C_O1_t[s,:,int(pair_all[iPair,2])]=C_O1_t[s,:,int(pair_all[iPair,2])]+t[1:3,0].T
            C_O2[s,:,:,int(pair_all[iPair,2]),int(pair_all[iPair,3])]=t[1:3,1:3]
            if s==iO:
                pair_all[iPair,1]=np.sum(t[1:3,1:3]*t[1:3,1:3])
        
        Ratio=np.sum(pair_all[:iPair,1])/(np.sum(pair_all[:iPair,0])+1e-10)
        Err_2nd=Ratio*np.sum(pair_all[iPair:,0])
        Pairs=np.hstack((Pairs,pair_all[iPair,2:4].reshape(-1,1)))  ##横向拼接，（2,1）与（2,1）的才可以拼接，与（2，）是拼接不起来的
    else:
        var=C_O2[iO,:,:,int(pair_all[iPair,2]),int(pair_all[iPair,3])]**2
        pair_all[iPair,1]=np.sum(var.reshape((4,1)))
        Ratio=np.sum(pair_all[:iPair,1])/(np.sum(pair_all[:iPair,0])+(1e-10))
        Err_2nd=Ratio*np.sum(pair_all[iPair:,0])
    return Err_2nd,C_O0_t,C_O1_t,Pairs,pair_all,Npair,C_O2
        


            
            
            
            